### PR TITLE
[CVM-1697] 2 Receiver error propagate to client and gateway

### PR DESCRIPTION
**Pull request by siscia**
Sup #2551 

Same PR but with changes in how we create JSON.

### CODE REVIEW

**Comment by siscia** [comment]
@cernvm-bot cloudtest

File: cvmfs/json_document_write.h [hunk 1]
```diff
@@ -14,34 +15,99 @@
...
 
-    JsonStringEntry() : is_quoted(true) {}
+    JsonEntry(std::string key, float val)
+        : variant(Float),
+          key(key),
+          str_val(),
+          int_val(0),
+          float_val(val) {}
+
+    JsonEntry(std::string key, int64_t val)
```
[the remark this answers is not in the document]
Comment (siscia) [reply cvmfs/json_document_write.h:48]: Yes and now, the `int` constructor is useful in calls like `input.Add("simple_number", 42)`, without the `int` constructor `int64_t` and `float` would be ambiguous.

File: cvmfs/json_document_write.h [hunk 2]
```diff
@@ -14,34 +15,99 @@
...
+          str_val(),
+          int_val(val),
+          float_val(0) {}
+
+    const std::string format() const {
+      char buffer[64];
+      int rc = -1;
+      switch (variant) {
+        case String:
+          return "\"" + key + "\":\"" + str_val + "\"";
```
[the remark this answers is not in the document]
Comment (siscia) [reply cvmfs/json_document_write.h:73]: Hummm, it is not up to the user, we escape the string. The user should pass the standard string. The escaping is done below in the Add methods.

Isn't better?

File: cvmfs/json_document_write.h [hunk 3]
```diff
@@ -14,34 +15,99 @@
...
+
+  void Add(std::string key, int64_t val) {
+    JsonEntry entry(Escape(key), val);
+    entries.push_back(entry);
+  }
+
+  void AddJsonObject(std::string key, std::string json) {
+    // we **do not escape** the value here
+    JsonEntry entry(Escape(key), json);
+    entry.variant = JsonObject;
```
[the remark this answers is not in the document]
Comment (siscia) [reply cvmfs/json_document_write.h]: I re-work a bit the code, maybe now it is cleaner and the comment superfluous, what you think?

File: cvmfs/json_document_write.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (jblomer) [reply cvmfs/json_document_write.h]: fair enough

File: cvmfs/json_document_write.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (jblomer) [reply cvmfs/json_document_write.h:78]: I think that's fine.  Perhaps let's rename the variables to `key_escaped` and `str_val_escaped`

**Pull request merged**